pkg/kv/kvclient/rangefeed/rangefeed.go

```diff
@@ -68,6 +68,41 @@ type Factory struct {
 	knobs   *TestingKnobs
 }
 
+// EventHandler is a rangefeed event handler. We use an interface such that
+// callers must explicitly handle each event type (if just to ignore them),
+// especially as new event types are added.
+type EventHandler interface {
+	// OnValue is called when the rangefeed emits a key/value write.
+	OnValue(context.Context, *roachpb.RangeFeedValue)
+
+	// OnCheckpoint is called when a rangefeed checkpoint occurs.
+	OnCheckpoint(context.Context, *roachpb.RangeFeedCheckpoint)
```

Comment:
One thing where this kind of interfaces looses out to callback based on is the fact that both OnValue
and OnCheckpoint could be applicable to both the initial scan and the rangefeed portions.

Certainly, OnValue is used in both, and I'm adding a checkpoint event when running initial scan.

I don't think that returning an error when encountering an event for which handler
was not configured is that bad, is it?
This API makes it a bit unwieldy...

Reply:
I'm not sure if I understand the difference? This API is functionally identical to the callback-based API. Both interfaces are intended to be implemented by the same struct (see examples), the only reason for having two interfaces instead of one is so that callers don't have to implement the initial scan methods unless they actually enable initial scans. We could choose to merge these interfaces if we wanted to.

>I don't think that returning an error when encountering an event for which handler was not configured is that bad, is it?

The advantage with this API is that we get errors at compile-time rather than at runtime. It's nice to avoid the possibility of rare race conditions ending up killing nodes or breaking components.